README.md
Get Started
=====

> 鉴于目前 ISSUE 较多而维护时间较少，且在进行后续的版本更新，目前暂时关闭 ISSUE，若社区有人跟进，欢迎和我们联系。重复的问题，请参阅常见问题的 [FAQ Wiki](https://github.com/fex-team/ueditor/wiki/FAQ)。

## ueditor富文本编辑器介绍

UEditor是由百度web前端研发部开发所见即所得富文本web编辑器，具有轻量，可定制，注重用户体验等特点，开源基于MIT协议，允许自由使用和修改代码。

## 1 入门部署和体验

### 1.1 下载编辑器

1. `git clone ` 仓库
2. `npm install` 安装依赖（如果没有安装 grunt , 请先在全局安装 grunt）
3. 在终端执行 `grunt default`

### 1.2 创建demo文件
解压下载的包，在解压后的目录创建demo.html文件，填入下面的html代码

```html
<!DOCTYPE HTML>
<html lang="en-US">
<head>
	<meta charset="UTF-8">
	<title>ueditor demo</title>
</head>
<body>
	<!-- 加载编辑器的容器 -->
	<script id="container" name="content" type="text/plain">这里写你的初始化内容</script>
	<!-- 配置文件 -->
	<script type="text/javascript" src="ueditor.config.js"></script>
	<!-- 编辑器源码文件 -->
	<script type="text/javascript" src="ueditor.all.js"></script>
	<!-- 实例化编辑器 -->
	<script type="text/javascript">
	    var ue = UE.getEditor('container');
	</script>
</body>
</html>
```

### 1.3 在浏览器打开demo.html

如果看到了下面这样的编辑器，恭喜你，初次部署成功！

![部署成功](http://fex.baidu.com/ueditor/doc/images/demo.png)

### 1.4 传入自定义的参数

编辑器有很多可自定义的参数项，在实例化的时候可以传入给编辑器：
```javascript
var ue = UE.getEditor('container', {
    autoHeight: false
});
```

配置项也可以通过ueditor.config.js文件修改，具体的配置方法请看[前端配置项说明](http://fex.baidu.com/ueditor/#start-config1.4 前端配置项说明.md)

### 1.5 设置和读取编辑器的内容

通getContent和setContent方法可以设置和读取编辑器的内容
```javascript
var ue = UE.getEditor();
//对编辑器的操作最好在编辑器ready之后再做
ue.ready(function(){
    //设置编辑器的内容
    ue.setContent('hello');
    //获取html内容，返回: <p>hello</p>
    var html = ue.getContent();
    //获取纯文本内容，返回: hello
    var txt = ue.getContentTxt();
});
```

ueditor的更多API请看[API 文档](http://ueditor.baidu.com/doc "ueditor API 文档")

## 2 详细文档

ueditor 官网：[http://ueditor.baidu.com](http://ueditor.baidu.com "ueditor 官网")

ueditor API 文档：[http://ueditor.baidu.com/doc](http://ueditor.baidu.com/doc "ueditor API 文档")

ueditor github 地址：[http://github.com/fex-team/ueditor](http://github.com/fex-team/ueditor "ueditor github 地址")

ueditor 第三方插件贡献 wiki : [第三方插件贡献规范](http://ueditor.baidu.com/website/thirdproject.html)

ueditor 贡献代码规范（javascript）： [javascript规范](https://github.com/fex-team/styleguide/blob/master/javascript.md)

## 3 第三方贡献

ueditor for nodejs 参考[https://github.com/netpi/ueditor](https://github.com/netpi/ueditor)

## 4 联系我们

email：[ueditor@baidu.com](mailto://email:ueditor@baidu.com "发邮件给ueditor开发组")

issue：[github issue](http://github.com/fex-team/ueditor/issues "ueditor 论坛")


changelog.md
#UEditor Change List
##1.5.0
###bug修复&优化改进
1.修复对嵌套的span标签设置样式时，被嵌套的标签不能正确设置上给定的样式

##1.4.4
###bug修复&优化改进
1. 修复展示页中li没有加前置选择器，可能导致影响展示页中li的css问题
2. 产出的数据中的color自动从rgb格式变成16进制格式
3. 过滤掉word中粘贴过来的a标签被误命中为锚点
4. 扩展的UI组件需要配置toolbars参数才可显示
5. ASP 创建目录 BUG 修复
6. JSP版本修复默认上传路径错误的bug。
7. 添加链接前缀的白名单，白名单里的前缀不加 http://

##1.4.3
###bug修复&优化改进
1. 修复hasContents接口在非ie下只有空格时判断还为真的问题
2. 修复在粘贴word内容时，会误命中cm,pt这样的文本内容变成px的问题
3. 优化删除编辑器再创建编辑器时，编辑器的容器id发生变化的问题
4. 修复提交jsonp请求时，callback参数的xss漏洞
5. 新增jsp后台多种服务器配置下的路径定位
6. 修复ZeroClipboard的flash地址参数名称错误
7. 修复getActionUrl的bug
8. 整理配置参数，把遗漏在代码中的配置项整理到ueditor.config.js里
9. 修复图片拉伸工具和编辑拉伸长高器的样式冲突
10. 修复文字的unicode编码会被错误再次解析问题
11. 添加消息提示功能，冒泡提示信息
12. 优化上传功能提示，当后端配置项没正常加载，禁用上传功能
13. 修复单图上传按钮和jqueryValidate不兼容的问题
14. 简化了与jqueryValidate的结合操作，具体看_examples/jqueryValidateDemo.html
15. 修复在删除编辑器后，再次创建时丢失原有id的问题
16. 修复查找替换在一些块节点中会导致替换错误

##1.4.2
###重构前后端交互功能
1. 前端上传模块统一改用webuploader
2. 整体重构了文件上传的配置方式，改为统一在后端配置，前后端自动打通，[详细文档]()
3. 统一各上传模块的提交地址，各模块通过action参数区分类型，[详细文档]()
4. 提供serverparam命令，可在提交时追加任意参数，[详细文档]()
5. 统一了前端各上传模块的布局样式
6. 支持了在线附件预览和插入
7. 统一了后端返回数据格式，[详细文档]()
8. 各在线预览列表模块支持分组加载
9. 增加点击直接选择文件上传图片插件
10. 优化了粘贴图片的功能，上传时有loading和出错的提示
11. 添加jsonp的跨域请求支持

###bug修复&优化改进
1. 修复内容过多时，回到顶部，顶部内容被工具栏遮挡问题
2. 修复htmlparser解析标签的bug
3. 修复锚点是#开头时还添加http://的bug
4. 修复全屏后，退出全屏高度没有缩回来的问题
5. 文字选中后按delete删除，无法触发contentchange事件
6. 修复选择图片时点击8个控制点不拖动，contentchange会误触发的问题
7. 修复执行命令会触发多次contentchange事件
8. 修复输入文字，设置高度300，没有滚动条
9. 修复在不可编辑模式下，链接认可修改的问题
10. 修复全局css对于ueditor有时展现会出现按钮独占一行的问题
11. 修复在ie11下上传图片失败的问题
12. 修复chrome 34版本下报错问题
13. 修复ie8下插入多张远程图片之后，对话框假死问题
14. 修复uparse,在页面中多次调用某些模块无效果问题
15. 修复容器宽度100%时,改变窗口大小,宽度不自适应的问题
16. 将桥接ui和编辑器的业务代码放到了新建的adapter目录，方便维护
17. 修复拖拽改变图片大小功能和bootstrap不兼容的问题
18. 修复在表格水平边框上拖拽，不能改变大小的问题
19. 修复在表格标题上可以向左向下合并单元格的问题
20. 修复grunt打包出错的问题
21. 优化ie11下的兼容问题，主要修复了表格下的诸多问题
22. 优化插入图片，添加原文件名作为alt属性

###新功能添加
1. 添加了enableContextMenu配置开关，开关右键菜单，默认为true
2. 添加disablePInList配置，指定产出的列表中是否嵌套P标签，默认是false
3. 添加retainOnlyLabelPasted配置，指定粘贴时是否是只保留标签模式，默认是false
4. 优化了添加toolbar上各类ui的方式，方便二次开发
6. 优化有时产出数据会带有 bookmark标签的脏数据问题
7. 添加LocalStorage本地存储工具
8. 优化自动排版功能，选项保存到localStorage或UserData
9. 添加右键菜单的复制粘贴的支持，非ie不提供粘贴功能

##1.3.6

###bug修复
1.	script/style标签内容，在ie下的编辑状态显示内容问题
2.	修复预览窗口没有滚动条问题 [出自](http://www.ueditorbbs.com/forum.php?mod=viewthread&tid=28231&extra=page%3D1)
3.	修复在ie67下自动寻址的问题
4.	修复ff下图片拖拽大小不触发contentchange问题
5.	修复注释被删除的问题，包括在script中的[出自](http://www.ueditorbbs.com/forum.php?mod=viewthread&tid=12509&extra=page%3D1)
6.	修复在源码模式下不能使用setContent的问题，[出自](http://www.ueditorbbs.com/forum.php?mod=viewthread&tid=26910&extra=page%3D1)
7.	修复在给定宽度为100%时，ie11在拖动窗口大小时，编辑区域不随着变化的问题
8.	修复在ie67下destroy方法调用报错
9.	修复在插入代码模式下，ie9+以上的浏览器键盘操作(enter/tab等)错误或者失效的问题
10.	修复不规则源码在ie下解析成文本的问题
11.	修复p标签在ie下嵌套的问题
12. 修复目录大纲更新事件导致失焦点的问题
13. 针对word粘贴列表到编辑器中自动转换为list标签，由于有误命中情况，所以默认关闭该功能。提供**autoTransWordToList**配置项(默认为false)，开发者可酌情配置。
14. 添加禁止表格嵌套的开关，解决excel中粘贴到编辑器中会有冗余的嵌套表格问题，配置项**disabledTableInTable**(默认是true)。
15. 过滤掉excel的表格粘贴到ie中时，会有bitmap的冗余占位图片问题

###功能更新
1.	支持视频上传插入，多浏览器自适配播放器

#1.3.5

###新增功能
1.	asp后台的支持
2.	添加本地自动保存功能
3.	增加数据可视化展示功能
4.	编辑器实例上添加isFocus,blur方法
5.	新增在chrome下针对图片的拖拽宽高功能
6.	新增在高端浏览器下，qq截图粘贴上传，拖拽图片到编辑上传
7.	添加表格插入列标题功能
8.	添加设置表格可排序功能，支持表格在预览页排序
9.  添加生成目录功能

###已有功能优化
1.	ie8以上版本使用w3cRange
2.	使用grunt作为打包工具
3.	修复了过滤规则对于script/style的内容的转码
5.	自动寻址功能重构
6.	修复下拉菜单高度问题
7.	针对ie默认带有的autolink功能，添加开启禁用选项，创建时传入autolink:false就可禁用ie的autolink功能
8.	支持插入动态地图
9.	**API文档更新**
10.	图片上传路径可配置，增加前后端路径验证
11.	对uparse进行了拆分重构
12.	随下载包提供各种功能说明文档
13. 背景颜色功能重构，可以在预览页显示背景
14. 重写了查找替换插件，解决ff下window.find方法失效的问题


##1.2.6.1
1.	查找替换支持正则表达式
2.	修复了ie67下初始化宽高给定百分比
3.	增加类似word中的快捷菜单，默认关闭
4.	针对默认过滤回转换div为p标签，提供了配置开关allowDivTransToP,默认为true
5.	修复了在ie下删除分割线后光标定位的问题
6.	提供了手动加载语言文件，避免ie下有时会因语言文件加载失败导致编辑器加载失败,提示"not import language file"的错误
7.	优化了编辑器初始化时获得contentWindow可能不存在的情况
8.	优化了编辑器加载自定义样式的问题，默认initialStyle传入的css样式优先级最高，其次是指定的外部css文件
9.	工具栏支持指定位置折行，'|'表示分割符,'||'表示折行
10.	表格操作功能升级，优化了对表格的拖拉及双击操作，并且支持IE6+浏览器。
11.	修复编辑器在禁用状态下仍然可以拖动表格边框的bug。
12.	修复了分割线不能删除的问题
13.	修复了初始化内容过多时，编辑器不自动长高，要点击编辑器才会长高的问题
14.	优化了添加字符边框的展示效果，避免出现重叠的问题
15.	修复下拉菜单超出屏幕的bug
16.	修复table属性初始化时table布局错误的bug
17.	优化了选择工具栏上下拉菜单类型的操作命令时，选区会有闪动的问题
18.	优化了关于swfupload的一个xss漏洞
19.	优化了对于ie9,10的支持


##1.2.6
1.	优化了编辑器路径的设置，可以不用手动设置路径，自动识别相关路径，解决路径设置繁琐的问题
2.	重写了过滤粘贴机制，采用黑白名单，可以书写符合自己需求的过滤规则，可以完全定义标签的属性，甚至是style上的某个属性及其数值
3.	优化了拖拽机制，处理浮动图片拖拽不能跟指定的某行对齐
4.	数据同步改为失去焦点就执行，可以不再使用sync方法手动同步数据
5.	添加了字体边框
6.	优化了backspace/del键的操作
7.	重写了插入代码功能，插入代码编写支持tab和回车键
8.	表格支持排序和隔行显示
9.	改使用closure的压缩工具
10.	优化了undo/redo操作
11.	优化了ui界面


##1.2.5
###新增功能
1.	table整体重构
2.	table支持插入表头和标题
3.	table支持拷贝
4.	table支持任意调整宽高
5.	table支持任意前插后插行列
6.	table键盘操作仿word用户体验
7.	添加table平均分布行、列
8.	添加table单元格对齐方式
9.	添加table对齐方式
10.	添加选中部分表格,点击backspace或delete删除功能
11.	重写表格属性、单元格属性dialog
12.	粘贴支持纯文本,源码,纯标签3个模式选择
13.	添加计算字数的getContentLength接口
14.	添加计算字数事件wordcount
15.	图片上传支持参数动态绑定
16.	重写了list功能，支持一,一),(一),1),(1),——等新的列表标签
17.	调整了list中tab键的逻辑
18.	添加了可以限制列表的层级
19.	全屏快捷键 ALT+Z
20.	添加了uparseplugin.js展示页加载器

###优化修复
1.	优化了插入代码功能
2.	ie下默认禁用源码模式下的代码高亮
3.	截图功能支持非ie浏览器
4.	修正了非ie下中文输入时回退不准确的问题
5.	改进了键盘输入时做回退的操作

##1.2.4
###新增功能
1.	官网新增API文档
2. CSS按照UI结构进行了模块化拆分
3. 新增皮肤切换功能，并提供一套新皮肤（可通过配置项theme来设置）
4. 新增编辑器容器拖动缩放功能,配置项为:scaleEnabled、minFrameWidth、minFrameHeight
5. 新增音乐插件
6. 增加了源码模式下,全屏按钮可以使用
7. 添加了UE.getEditor工厂方法
8. 添加了针对jquery配合使用的demo
9. 添加了针对jqueryValidation配合使用的demo
10. 添加了初始化编辑器宽高配置，配置为项:initialFrameWidth、initialFrameHeight

###优化修复
1. 修复涂鸦路径在配置时，添加参数时请求报错
2. 修复涂鸦opera下缩放不能使用
3. 修复编辑器全屏功能失效问题
4. codemirror版本升级到最新版
5. 对opera/safari的支持进行了进一步的优化
6. 优化了部分demo页的代码
7. 修改原来的minFrameHeight为拖动时的最小高度

##1.2.3
###新增功能
1. 新增国际化支持
2. 新增涂鸦功能
3. 新增大小写功能
4. 新增getAllHtml方法，可以将整个页面的内容打出来,可以在ueditor.configplugin.js里通过配置allHtmlEnabled,来配置在提交时是否使用getAllHtml来得到内容
5. 新增插入模板的功能
6. 新增背景功能
7. 新增UE.instants全局对象，下边挂接了所有实例化的编辑器对象
8. Editor下新增ready方法，当编辑器ready后执行传入的fn,如果编辑器已经ready好了，就马上执行fn
9. 新增topOffset配置参数，用于设置AutoFloat时工具栏距离顶部的高度
10. 新增sourceEditorFirst配置参数，用于控制编辑器初始化时是否显示成源码模式，默认为否
11. 新增在表格内实例化编辑器的demo
12. 新增getDialog(dialogName)接口，可以获取dialog对象。

###优化修复

js/list.js
var docList = [
    {
        id: 'start',
        title: '1. 开始使用',
        list: [
            {
                id:'start',
                title: '1.1 开始使用ueditor'
            },
            {
                id:'toolbar',
                title: '1.2 定制工具栏图标'
            },
            {
                id: 'config',
                title: '1.3 前端配置项说明'
            },
            {
                id: 'dir',
                title: '1.4 目录介绍'
            },
            {
                id: 'submit',
                title: '1.5 提交表单'
            },
            {
                id: 'uparse',
                title: '1.6 编辑内容展示'
            },
            {
                id: 'patch',
                title: '1.7 如何使用补丁文件'
            }//,
//            {
//                id: 'shortkey',
//                title: '1.8 设置快捷键'
//            },
//            {
//                id: 'event',
//                title: '1.9 监听编辑器事件'
//            },
//            {
//                id: 'filter',
//                title: '1.10 粘贴过滤器(黑白名单)'
//            }
        ]
    },
    {
        id: 'server',
        title: '2. 后端配置',
        list: [
            {
                id:'deploy',
                title: '2.1 后端部署说明'
            },
            {
                id:'php',
                title: '2.2 PHP 使用说明'
            },
            {
                id:'asp',
                title: '2.3 ASP 使用说明'
            },
            {
                id:'net',
                title: '2.4 ASP.NET 使用说明'
            },
            {
                id:'jsp',
                title: '2.5 JSP 使用说明'
            },
            // {
            //     id: 'django',
            //     title: '2.6 Python Django 使用说明'
            // },
            {
                id:'config',
                title: '2.7 后端配置项说明'
            },
            {
                id:'path',
                title: '2.8 上传路径配置'
            },
            {
                id: 'video',
                title: '2.9 上传视频'
            }
        ]
    },
    {
        id: 'dev',
        title: '3. 二次开发',
        list: [
            {
                id: 'developer',
                title: '3.1 如何进行二次开发'
            },
            {
                id: 'bale_width_grunt',
                title: '3.2 使用grunt打包源代码'
            },
            {
                id:'request_specification',
                title: '3.3 后端请求规范'
            },
            {
                id:'serverparam',
                title: '3.4 自定义请求参数'
            },
            {
                id:'crossdomain',
                title: '3.5 跨域支持说明'
            }//,
//            {
//                id: '',
//                title: '3.6 修改编辑区域样式'
//            },
//            {
//                id: '',
//                title: '3.7 输入输出过滤规则'
//            },
        ]
    },
    {
        id: 'api',
        title: '4. 常用API列表',
        list: [
            {
                id: 'common',
                title: '4.1 常用API'
            }
        ]
    },
    {
        id: 'qa',
        title: '5. 常见问题',
        list: [
            {
                id:'ready',
                title:'5.1 实例创建后为什么直接执行接口报错'
            },
            {
                id:'customurl',
                title:'5.2 如何自定义请求地址'
            },
            {
                id:'allowDivToP',
                title:'5.3 如何阻止div变成p标签'
            }

        ]
    }
];
